File: src/data/courses/dsa/trees/structural-property-problems/diameter-of-tree/code/java/DiameterOfTree.java
// ============================================================
// Diameter of Binary Tree
// ============================================================

class Solution {
    int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        diameter = 0;
        height(root);
        return diameter;
    }
    int height(TreeNode node) {
        if (node == null) return 0;
        int l = height(node.left), r = height(node.right);
        diameter = Math.max(diameter, l + r);
        return 1 + Math.max(l, r);
    }
}

File: src/data/courses/dsa/trees/structural-property-problems/diameter-of-tree/code/python/diameter_of_tree.py
# ============================================================
# Diameter of Binary Tree
# ============================================================

class Solution:
    def diameterOfBinaryTree(self, root) -> int:
        self.diameter = 0

        def height(node):
            if not node: return 0
            left_h = height(node.left)
            right_h = height(node.right)
            # Diameter through this node = left + right
            self.diameter = max(self.diameter, left_h + right_h)
            return 1 + max(left_h, right_h)

        height(root)
        return self.diameter
